src/theory/quantifiers/term_util.cpp

```diff
@@ -690,43 +693,31 @@ Node TermUtil::ensureType( Node n, TypeNode tn ) {
   }
 }
 
-bool TermUtil::containsTerm2( Node n, Node t, std::map< Node, bool >& visited ) {
-  if( n==t ){
-    return true;
-  }else{
-    if( visited.find( n )==visited.end() ){
-      visited[n] = true;
-      for( unsigned i=0; i<n.getNumChildren(); i++ ){
-        if( containsTerm2( n[i], t, visited ) ){
-          return true;
-        }
-      }
-    }
-    return false;
-  }
-}
-
 bool TermUtil::containsTerms2( Node n, std::vector< Node >& t, std::map< Node, bool >& visited ) {
-  if( visited.find( n )==visited.end() ){
+  if (visited.find(n) == visited.end())
+  {
     if( std::find( t.begin(), t.end(), n )!=t.end() ){
       return true;
-    }else{
-      visited[n] = true;
-      for( unsigned i=0; i<n.getNumChildren(); i++ ){
-        if( containsTerms2( n[i], t, visited ) ){
-          return true;
-        }
+    }
+    visited[n] = true;
+    if (n.hasOperator())
+    {
+      if (containsTerms2(n.getOperator(), t, visited))
```

Comment:
merge ``if``

Reply:
i dont like merging ifs if things become sensitive to evaluation order.  Here, if !n.hasOperator(), then calling containsTerms2(n.getOperator()...) will crash.

Reply:
but if ``!n.hasOperator()`` the function would not be called. I don't think separating the ``if`` helps any more than writing a single one

Reply:
If they are merged:

if( n.hasOperator() && containsTerms2(n.getOperator(), t, visited))

if the second conjunct is evaluated first, the function is called and CVC4 may crash.  Generally, I don't like relying on the evaluation order of && being strict.

Reply:
I think it's fair to assume that && will always be evaluated strictly in this order.

Reply:
I disagree, I have been bitten by this in the past.  Plus I think the code is better documented if these are kept separate: this is saying that the developer wants to enforce the evaluation order.